File: android/src/main/java/name/zeno/android/behavior/VerticalScrollingBehavior.java
package name.zeno.android.behavior;

import android.content.Context;
import android.support.annotation.IntDef;
import android.support.design.widget.CoordinatorLayout;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.View;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * <h1>垂直滚动</h1>
 * <p>
 * 主要的方法
 * <ul>
 * <li>{@link #onDirectionNestedPreScroll(CoordinatorLayout, View, View, int, int, int[], int)}</li>
 * <li>{@link #onNestedVerticalOverScroll(CoordinatorLayout, View, int, int, int)}</li>
 * <li>{@link #onNestedDirectionFling(CoordinatorLayout, View, View, float, float, int)}</li>
 * </ul>
 *
 * @author 陈治谋 (513500085@qq.com)
 * @since 16/8/22
 */
@SuppressWarnings("unused")
public abstract class VerticalScrollingBehavior<V extends View> extends CoordinatorLayout.Behavior<V>
{

  private static final String TAG = "VerticalScrollingBehavior";

  private int mTotalDyUnconsumed = 0;
  private int mTotalDy           = 0;

  @ScrollDirection
  private int mOverScrollDirection = ScrollDirection.SCROLL_NONE;
  @ScrollDirection
  private int mScrollDirection     = ScrollDirection.SCROLL_NONE;

  public VerticalScrollingBehavior()
  {
    super();
  }

  public VerticalScrollingBehavior(Context context, AttributeSet attrs)
  {
    super(context, attrs);
  }

  public int getMOverScrollDirection()
  {return this.mOverScrollDirection;}

  public int getMScrollDirection()
  {return this.mScrollDirection;}

  @Retention(RetentionPolicy.SOURCE)
  @IntDef({ScrollDirection.SCROLL_DIRECTION_UP, ScrollDirection.SCROLL_DIRECTION_DOWN})
  public @interface ScrollDirection
  {
    int SCROLL_DIRECTION_UP   = 1;
    int SCROLL_DIRECTION_DOWN = -1;
    int SCROLL_NONE           = 0;
  }

  @Override
  public boolean onStartNestedScroll(CoordinatorLayout coordinatorLayout, V child, View directTargetChild, View target, int nestedScrollAxes)
  {
    return (nestedScrollAxes & ViewCompat.SCROLL_AXIS_VERTICAL) != 0;
  }

  @Override
  public void onNestedPreScroll(CoordinatorLayout coordinatorLayout, V child, View target, int dx, int dy, int[] consumed)
  {
    super.onNestedPreScroll(coordinatorLayout, child, target, dx, dy, consumed);
    if (dy > 0 && mTotalDy < 0) {
      mTotalDy = 0;
      mScrollDirection = ScrollDirection.SCROLL_DIRECTION_UP;
    } else if (dy < 0 && mTotalDy > 0) {
      mTotalDy = 0;
      mScrollDirection = ScrollDirection.SCROLL_DIRECTION_DOWN;
    }
    mTotalDy += dy;
    onDirectionNestedPreScroll(coordinatorLayout, child, target, dx, dy, consumed, mScrollDirection);
  }

  @Override
  public void onNestedScroll(CoordinatorLayout coordinatorLayout, V child, View target, int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed)
  {
    super.onNestedScroll(coordinatorLayout, child, target, dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed);
    if (dyUnconsumed > 0 && mTotalDyUnconsumed < 0) {
      mTotalDyUnconsumed = 0;
      mOverScrollDirection = ScrollDirection.SCROLL_DIRECTION_UP;
    } else if (dyUnconsumed < 0 && mTotalDyUnconsumed > 0) {
      mTotalDyUnconsumed = 0;
      mOverScrollDirection = ScrollDirection.SCROLL_DIRECTION_DOWN;
    }
    mTotalDyUnconsumed += dyUnconsumed;
    onNestedVerticalOverScroll(coordinatorLayout, child, mOverScrollDirection, dyConsumed, mTotalDyUnconsumed);
  }

  @Override
  public boolean onNestedFling(CoordinatorLayout coordinatorLayout, V child, View target, float velocityX, float velocityY, boolean consumed)
  {
    super.onNestedFling(coordinatorLayout, child, target, velocityX, velocityY, consumed);
    mScrollDirection = velocityY > 0 ? ScrollDirection.SCROLL_DIRECTION_UP : ScrollDirection.SCROLL_DIRECTION_DOWN;
    return onNestedDirectionFling(coordinatorLayout, child, target, velocityX, velocityY, mScrollDirection);
  }

  /**
   * @param direction         {@link ScrollDirection}
   * @param currentOverScroll Unconsumed value, negative or positive based on the direction;
   * @param totalOverScroll   Cumulative value for current direction
   */
  public void onNestedVerticalOverScroll(CoordinatorLayout coordinatorLayout, V child, @ScrollDirection int direction, int currentOverScroll, int totalOverScroll) {}

  /**
   * @param scrollDirection {@link ScrollDirection}
   */
  public void onDirectionNestedPreScroll(CoordinatorLayout coordinatorLayout, V child, View target, int dx, int dy, int[] consumed, @ScrollDirection int scrollDirection) {}

  /**
   * @return true if the Behavior consumed the fling
   */
  protected boolean onNestedDirectionFling(CoordinatorLayout coordinatorLayout, V child, View target, float velocityX, float velocityY, @ScrollDirection int scrollDirection)
  {
    return false;
  }

}

File: library/src/main/java/name/zeno/android/behavior/VerticalScrollingBehavior.kt
package name.zeno.android.behavior

import android.content.Context
import android.support.annotation.IntDef
import android.support.design.widget.CoordinatorLayout
import android.support.v4.view.ViewCompat
import android.util.AttributeSet
import android.view.View

/**
 * <h1>垂直滚动</h1>
 *
 *
 * 主要的方法
 *
 *  * [.onDirectionNestedPreScroll]
 *  * [.onNestedVerticalOverScroll]
 *  * [.onNestedDirectionFling]
 *
 *
 * @author 陈治谋 (513500085@qq.com)
 * @since 16/8/22
 */
abstract class VerticalScrollingBehavior<V : View> : CoordinatorLayout.Behavior<V> {

  private var mTotalDyUnconsumed = 0
  private var mTotalDy = 0

  @ScrollDirection
  var mOverScrollDirection = SCROLL_NONE
    private set
  @ScrollDirection
  var mScrollDirection = SCROLL_NONE
    private set

  constructor() : super() {}

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

  @IntDef(SCROLL_DIRECTION_UP, SCROLL_DIRECTION_DOWN)
  @Retention(AnnotationRetention.SOURCE)
  annotation class ScrollDirection()

  override fun onStartNestedScroll(coordinatorLayout: CoordinatorLayout, child: V, directTargetChild: View, target: View, nestedScrollAxes: Int): Boolean {
    return nestedScrollAxes and ViewCompat.SCROLL_AXIS_VERTICAL != 0
  }

  override fun onNestedPreScroll(coordinatorLayout: CoordinatorLayout, child: V, target: View, dx: Int, dy: Int, consumed: IntArray) {
    super.onNestedPreScroll(coordinatorLayout, child, target, dx, dy, consumed)
    if (dy > 0 && mTotalDy < 0) {
      mTotalDy = 0
      mScrollDirection = SCROLL_DIRECTION_UP
    } else if (dy < 0 && mTotalDy > 0) {
      mTotalDy = 0
      mScrollDirection = SCROLL_DIRECTION_DOWN
    }
    mTotalDy += dy
    onDirectionNestedPreScroll(coordinatorLayout, child, target, dx, dy, consumed, mScrollDirection)
  }

  override fun onNestedScroll(coordinatorLayout: CoordinatorLayout, child: V, target: View, dxConsumed: Int, dyConsumed: Int, dxUnconsumed: Int, dyUnconsumed: Int) {
    super.onNestedScroll(coordinatorLayout, child, target, dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed)
    if (dyUnconsumed > 0 && mTotalDyUnconsumed < 0) {
      mTotalDyUnconsumed = 0
      mOverScrollDirection = SCROLL_DIRECTION_UP
    } else if (dyUnconsumed < 0 && mTotalDyUnconsumed > 0) {
      mTotalDyUnconsumed = 0
      mOverScrollDirection = SCROLL_DIRECTION_DOWN
    }
    mTotalDyUnconsumed += dyUnconsumed
    onNestedVerticalOverScroll(coordinatorLayout, child, mOverScrollDirection, dyConsumed, mTotalDyUnconsumed)
  }

  override fun onNestedFling(coordinatorLayout: CoordinatorLayout, child: V, target: View, velocityX: Float, velocityY: Float, consumed: Boolean): Boolean {
    super.onNestedFling(coordinatorLayout, child, target, velocityX, velocityY, consumed)
    mScrollDirection = if (velocityY > 0) SCROLL_DIRECTION_UP else SCROLL_DIRECTION_DOWN
    return onNestedDirectionFling(coordinatorLayout, child, target, velocityX, velocityY, mScrollDirection)
  }

  /**
   * @param direction         [ScrollDirection]
   * @param currentOverScroll Unconsumed value, negative or positive based on the direction;
   * @param totalOverScroll   Cumulative value for current direction
   */
  fun onNestedVerticalOverScroll(coordinatorLayout: CoordinatorLayout, child: V, @ScrollDirection direction: Int, currentOverScroll: Int, totalOverScroll: Int) {}

  /**
   * @param scrollDirection [ScrollDirection]
   */
  open fun onDirectionNestedPreScroll(coordinatorLayout: CoordinatorLayout, child: V, target: View, dx: Int, dy: Int, consumed: IntArray, @ScrollDirection scrollDirection: Int) {}

  /**
   * @return true if the Behavior consumed the fling
   */
  protected open fun onNestedDirectionFling(coordinatorLayout: CoordinatorLayout, child: V, target: View, velocityX: Float, velocityY: Float, @ScrollDirection scrollDirection: Int): Boolean {
    return false
  }

  companion object {
    const val SCROLL_DIRECTION_UP = 1
    const val SCROLL_DIRECTION_DOWN = -1
    const val SCROLL_NONE = 0
  }

}

